llvm/lib/IR/Value.cpp

```diff
@@ -909,7 +909,7 @@ uint64_t Value::getPointerDereferenceableBytes(const DataLayout &DL,
     if (!AI->isArrayAllocation()) {
       DerefBytes =
           DL.getTypeStoreSize(AI->getAllocatedType()).getKnownMinValue();
-      CanBeNull = false;
+      CanBeNull = AI->getType()->getPointerAddressSpace() != 0;
```

Comment:
I don't think this PR should be part of this change. Note that there is also a similar case for globals below.

I'm also not entirely sure this change should be made at all. There is a subtle distinction in semantics between:

 * Can return an allocation or null. This means that the pointer is *not* always dereferenceable.
 * Can return an allocation that may also be located at the null address. This means that the pointer *is* always dereferenceable, even if it is null.

The doc comment for this method is phrased vaguely enough that it's not super clear what the intended semantics are:

```
  /// If CanBeNull is set by this function the pointer can either be null or be
  /// dereferenceable up to the returned number of bytes.
```

But there are clearly some uses of the API where the current implementation is fine, for example
https://github.com/llvm/llvm-project/blob/9d364286f3b63e99ed3838f179aa2223f930f1ab/llvm/lib/Transforms/InstCombine/InstructionCombining.cpp#L3060 doesn't care whether the pointer is null, as long as it is dereferenceable.

The use here *does* care. I think what I would go for is to leave this method alone and drop the `!CanBeNull ||` check (maybe with a comment on why we don't use it).

Reply:
Fair enough and thanks for explaining the subtle distinction! I've reverted the changes.